File: nxplx-backend/NxPlx/EpisodeRoutes.fs
module NxPlx.EpisodeRoutes

open System
open NxPlx.Types
open NxPlx.TMdbApi
open NxPlx.Indexing
open Red
open Red.Interfaces

type EpisodeInfo = {id:int; eid:int; title:string; season:int; number:int; subtitles:string array; backdrop:string}

let register (router:IRouter) (indexer:Indexer) =        
        
    router.Get("/:id", fun (req:Request) (res:Response) ->
        let id = req.Context.ExtractUrlParameter "id" |> Int32.Parse
        let episode = Map.find id indexer.episodes
        let series = Map.find episode.series indexer.series
        res.SendJson {id=episode.id; eid=episode.eid; title=series.name; season=episode.season; number=episode.number;
                      subtitles=episode.subtitles; backdrop=episode.thumbnail})
          
    router.Get("/:eid/watch", fun (req:Request) (res:Response) ->
        let eid = req.Context.ExtractUrlParameter "eid" |> Int32.Parse
        let episode = Map.find eid indexer.episodeEntries
        res.SendFile (episode.path, "video/mp4", true, 1 * 1024 * 1024))

File: nxplx-backend/NxPlx.WebApi/Routers/EpisodeRoutes.cs
using System;
using System.IO;
using System.Linq;
using System.Net;
using Microsoft.EntityFrameworkCore;
using NxPlx.Models;
using NxPlx.Models.File;
using NxPlx.Services.Database;
using Red.Interfaces;

namespace NxPlx.WebApi.Routers
{
    public static class EpisodeRoutes
    {
        public static void Register(IRouter router)
        {
            router.Get("/:uuid", async (req, res) =>
            {
                var uuid = Guid.Parse(req.Context.ExtractUrlParameter("uuid"));
                using (var ctx = new MediaContext())
                {
                    var episode = await ctx.Episodes
                        .Include(e => e.Season)
                        .ThenInclude(s => s.Series)
                        .FirstOrDefaultAsync(e => e.Id == uuid);
                    
                    if (episode == null)
                    {
                        return await res.SendStatus(HttpStatusCode.NotFound);
                    }
                    
                    return await res.SendJson(episode);
                }
            });
            
            router.Get("/:uuid/details", async (req, res) =>
            {
                var uuid = Guid.Parse(req.Context.ExtractUrlParameter("uuid"));
                using (var ctx = new MediaContext())
                {
                    var episodeFile = await ctx.EpisodeFiles
                        .Include(e => e.Subtitles)
                        .Include(e => e.MediaDetails)
                        .FirstOrDefaultAsync(e => e.Id == uuid);

                    var episode = await ctx.Episodes.FindAsync(uuid);
                    
                    if (episodeFile == null)
                    {
                        return await res.SendStatus(HttpStatusCode.NotFound);
                    }
                    
                    return await res.SendJson(episodeFile);
                }
            });
            
            router.Get("/:uuid/watch", async (req, res) =>
            {
                var uuid = Guid.Parse(req.Context.ExtractUrlParameter("uuid"));
                using (var ctx = new MediaContext())
                {
                    var episode = await ctx.EpisodeFiles.FindAsync(uuid);
                    if (episode == null)
                    {
                        return await res.SendStatus(HttpStatusCode.NotFound);
                    }

                    using (var mediaStream = File.Open(episode.Path, FileMode.Open, FileAccess.Read, FileShare.Read))
                    {
                        return await res.SendStream(mediaStream, "video/mp4");
                    }
                }
            });
            
//            router.Get("/:eid/watch", fun (req:Request) (res:Response) ->
//                let eid = req.Context.ExtractUrlParameter "eid" |> Int32.Parse
//            let episode = Map.find eid indexer.episodeEntries
//            res.SendFile (episode.path, "video/mp4", true))
        }
    }
}